### PR TITLE
[Heartbeat] Fix timing for mode all monitors + Refactor

### PR DESCRIPTION
**Pull request by andrewvc**
Timing broke in 6.x in elastic/beats#9697 when `mode: all` is used. Some events will wind up with an empty timestamp, which in go is the year 0001. This isn't slated for release till 6.7 so there's no huge rush here.

This PR seeks to fix not only the bug, but the structural problems in the code that led to the bug. The key issue here is that the timing code should have been centralized for all beats, and invoked in one spot, vs. needing to be invoked in each different dialer type.

That is not a small refactor, but this PR has been able to achieve this and as a result make the code much easier to reason about, and thus safer.

A key thing here was handling the recursive nature of heartbeat tasks in a more uniform way. Previously some functions that modified jobs were recursive, other times they were not. Now, recursive application of modifications only happens in one spot the `wrapper.WrapAll` and `wrapper.Wrap` functions (the first one takes a list of jobs, while the second takes a single Job). A new type `JobWrapper`, can be passed into these. These Job wrapping functions  don't handle any recursive logic themselves. 

This PR also breaks up some of this utility code, moving it into separate packages. The `monitor` package became too unwieldy with the additions here.

More tests have been added. There is more work to do to clean stuff up, but there's more than enough here for a single PR.

### CODE REVIEW

**Comment by ruflin** [comment]
Trying to get my head around why the introduction of `mode: all` caused this issue and why this PR solves it. Could you share a bit more details?

**Comment by andrewvc** [comment]
@ruflin  the basic issue is that the original code did the timing deep in the logic in various different places. The refactor that broke things centralized the logic into a Job wrapper, but didn't apply it in all the right places. This fix moves the wrapper up to the very top level of job invocation so that all jobs are wrapped in it at all times.

The code's kind of hard to follow, but if you trace it you'll note that the `all` path (`makeByHostAllIPJob`) doesn't invoke `MakeByIPJob`, which is the only place we execute the timing code at the moment.

**Comment by andrewvc** [comment]
@ruflin I'll add the logic here is much clearer now, by wrapping jobs at the top level it's very clear that any job (and its descendants) will have timing applied to it.

**Comment by ruflin** [comment]
Thanks for the details, make sense now. It seems this PR breaks heartbeat CI? Could you also rebase to to get the other beats back to green?

File: heartbeat/monitors/active/http/http_test.go [hunk 1]
```diff
@@ -65,7 +65,7 @@ func testTLSRequest(t *testing.T, testURL string, extraConfig map[string]interfa
 	jobs, endpoints, err := create("tls", config)
 	require.NoError(t, err)
 
-	job := jobs[0]
+	job := monitors.WrapCommon(jobs, "tls", "", "http")[0]
```
Comment (andrewvc) [inline heartbeat/monitors/active/http/http_test.go]: Ideally we wouldn't have this in tests, and the call to `create` above would be a different call that would do the wrapping. Unfortunately `newMonitor`, where this happens now would need a refactor for this to be clean. This PR has enough scope, so I'd like to defer that refactor.

File: heartbeat/monitors/monitor.go [hunk 2]
```diff
@@ -121,44 +125,29 @@ func newMonitor(
...
-				job,
-			)
-		}
+		m.id = fmt.Sprintf("auto-%s-%#X", m.typ, hash)
 	}
 
-	jobs = WrapAll(jobs, idWrapper)
-
+	rawJobs, endpoints, err := monitorPlugin.create(config)
+	wrappedJobs := wrappers.WrapCommon(rawJobs, m.id, m.name, m.typ)
```
Comment (andrewvc) [inline heartbeat/monitors/monitor.go]: This is the key part of this PR. It is the new point at which all common fields are applied, including timing.

File: heartbeat/monitors/util.go [hunk 3]
```diff
@@ -261,21 +219,19 @@ func makeByHostAnyIPJob(
 		resolveRTT := resolveEnd.Sub(resolveStart)
 
 		ipFields := resolveIPEvent(ip.String(), resolveRTT)
-		return WithFields(ipFields, pingFactory(ip))(event)
+		return wrappers.WithFields(ipFields, pingFactory(ip))(event)
 	}
-
-	return TimeAndCheckJob(aj)
```
Comment (andrewvc) [inline heartbeat/monitors/util.go]: No longer necessary since this happens at the root invocation of all monitors!

File: heartbeat/monitors/util.go [hunk 4]
```diff
@@ -298,11 +254,11 @@ func makeByHostAllIPJob(
 		}
 
 		// create ip ping tasks
-		cont := make([]Job, len(ips))
+		cont := make([]jobs.Job, len(ips))
 		for i, ip := range ips {
 			addr := &net.IPAddr{IP: ip}
 			ipFields := resolveIPEvent(ip.String(), resolveRTT)
-			cont[i] = TimeAndCheckJob(WithFields(ipFields, pingFactory(addr)))
```
Comment (andrewvc) [inline heartbeat/monitors/util.go]: Also no longer necessary!

File: heartbeat/hbtest/hbtestutil.go [hunk 5]
```diff
@@ -97,13 +97,15 @@ func TLSChecks(chainIndex, certIndex int, certificate *x509.Certificate) mapval.
...
+func BaseChecks(ip string, status string, typ string) mapval.Validator {
 	return mapval.MustCompile(mapval.Map{
 		"monitor": mapval.Map{
-			"duration.us": mapval.IsDuration,
 			"ip":          ip,
+			"duration.us": mapval.IsDuration,
 			"status":      status,
+			"id":          mapval.IsNonEmptyString,
+			"name":        mapval.IsString,
+			"type":        typ,
```
Comment (andrewvc) [inline heartbeat/hbtest/hbtestutil.go]: Previously these checks were done when the event was incomplete. With the refactor we can test all fields.

File: heartbeat/eventext/eventext.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+)
+
+// MergeEventFields merges the given common.MapStr into the given Event's Fields.
+func MergeEventFields(e *beat.Event, merge common.MapStr) {
+	if e.Fields != nil {
+		e.Fields.DeepUpdate(merge)
+	} else {
+		e.Fields = merge
+	}
+}
```
Comment (urso) [inline heartbeat/eventext/eventext.go]: I didn't check if that's the case, but in general be always cautious when sharing `common.MapStr` via DeepUpdate or assigning to e.Fields. They might be modified by processors, potentially leading to panics and crashes due to races. If processors only apply to a subset of events you might also end up with invalid event fields. When in doubt call `MergeEventFields(e, fields.Clone())`.

Comment (urso) [reply heartbeat/eventext/eventext.go]: Btw. I'm aware you moved it :) Just wanted to point a common programming error we've seen in the past in heartbeat and a few other places. Maybe we should add a code comment telling users of this function to use this only with freshly constructed MapStrs or a Clone of an existing MapStr.

File: libbeat/common/mapval/is_defs.go [hunk 7]
```diff
@@ -192,6 +192,16 @@ func isStrCheck(path path, v interface{}) (str string, errorResults *Results) {
 	return strV, nil
 }
 
+// IsString checks that the given value is a string.
+var IsString = Is("is a non-empty string", func(path path, v interface{}) *Results {
```
Comment (urso) [inline libbeat/common/mapval/is_defs.go]: message: "is a string type"?

The `non-empty` label is already in use.

File: heartbeat/eventext/eventext.go (hunk 6 above)
Comment (andrewvc) [reply heartbeat/eventext/eventext.go]: Yes, I've checked this in the past. Ideally we'd enforce this via code instead of convention, but I'm not worried about clones at the moment.

File: heartbeat/monitors/active/dialchain/builder.go [hunk 8]
```diff
@@ -23,6 +23,7 @@ import (
 	"time"
 
 	"github.com/elastic/beats/heartbeat/monitors"
+	"github.com/elastic/beats/heartbeat/monitors/jobs"
```
Comment (ruflin) [inline heartbeat/monitors/active/dialchain/builder.go]: I'm wondering why the package is not call `job` instead of `jobs`?

File: heartbeat/monitors/jobs/testing.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+package jobs
+
+import (
+	"testing"
+
+	"github.com/elastic/beats/libbeat/beat"
+)
+
+// ExecJobsAndConts recursively executes multiple jobs.
+func ExecJobsAndConts(t *testing.T, jobs []Job) ([]*beat.Event, error) {
```
Comment (ruflin) [inline heartbeat/monitors/jobs/testing.go]: Conts?

File: heartbeat/monitors/wrappers/monitors.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+func WrapCommon(js []jobs.Job, id string, name string, typ string) []jobs.Job {
+	return jobs.WrapAll(
+		js,
+		monStatus,
+		monTiming,
+		monMeta(id, name, typ),
+	)
+}
+
+func monMeta(id string, name string, typ string) jobs.JobWrapper {
```
Comment (ruflin) [inline heartbeat/monitors/wrappers/monitors.go]: mon = monitor? I'm personally not a big fan of too many abbreviations as when reading the code I first always need to figure out what it could mean.

File: heartbeat/monitors/wrappers/monitors.go [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+			fields["error"] = look.Reason(err)
+		}
+		eventext.MergeEventFields(event, fields)
+		return cont, nil
+	}
+}
+
+// monTiming executes the given Job, checking the duration of its run and setting
+// its monStatus.
+// It adds the monitor.duration and monitor.monStatus fields.
```
Comment (ruflin) [inline heartbeat/monitors/wrappers/monitors.go]: Does it add the status?

Comment (andrewvc) [reply heartbeat/monitors/wrappers/monitors.go]: Ah, not anymore! I refactored that into `monStatus`. I'll fix the comment. They used to be combined.

File: heartbeat/monitors/wrappers/monitors.go (hunk 10 above)
Comment (andrewvc) [reply heartbeat/monitors/wrappers/monitors.go]: @ruflin I'm glad to extend it, but I thought inscrutable short names was the go way 😁 !

File: heartbeat/monitors/jobs/testing.go (hunk 9 above)
Comment (andrewvc) [reply heartbeat/monitors/jobs/testing.go]: Short for continuations.

File: heartbeat/monitors/active/dialchain/builder.go (hunk 8 above)
Comment (andrewvc) [reply heartbeat/monitors/active/dialchain/builder.go]: @ruflin the lame answer here is that it makes the code elsewhere confusings because we use the variable `jobs` a lot. I'm glad to change it, I guess elsewhere I could use `js := jobs.Job[]` instead of `jobs := job.Job[]`
